optimizely/decision_service.py

```diff
@@ -323,128 +329,216 @@ def get_variation(
+                decide_reasons.append(message)
 
-    def get_variation_for_feature(self, project_config, feature, user_id, attributes=None, ignore_user_profile=False):
+                # skip the rest of rollout rules to the everyone-else rule if audience matches but not bucketed.
+                skip_to_everyone_else = True
+
+        else:
+            message = 'User "{}" does not meet audience conditions for targeting rule {}.'.format(user_id, logging_key)
+            self.logger.debug(message)
+            decide_reasons.append(message)
+
+        return (bucketed_variation, skip_to_everyone_else), decide_reasons
+
+    def get_variation_for_feature(self, project_config, feature, user_context, ignore_user_profile=False):
         """ Returns the experiment/variation the user is bucketed in for the given feature.
 
     Args:
       project_config: Instance of ProjectConfig.
       feature: Feature for which we are determining if it is enabled or not for the given user.
-      user_id: ID for user.
+      user: user context for user.
       attributes: Dict representing user attributes.
       ignore_user_profile: True if we should bypass the user profile service
 
     Returns:
       Decision namedtuple consisting of experiment and variation for the user.
     """
         decide_reasons = []
-        bucketing_id, reasons = self._get_bucketing_id(user_id, attributes)
-        decide_reasons += reasons
 
         # Check if the feature flag is under an experiment and the the user is bucketed into one of these experiments
         if feature.experimentIds:
             # Evaluate each experiment ID and return the first bucketed experiment variation
             for experiment in feature.experimentIds:
                 experiment = project_config.get_experiment_from_id(experiment)
                 if experiment:
-                    variation, variation_reasons = self.get_variation(
-                        project_config, experiment, user_id, attributes, ignore_user_profile)
+                    variation, variation_reasons = self.get_variation_from_experiment_rule(
+                        project_config, feature.key, experiment, user_context, ignore_user_profile)
                     decide_reasons += variation_reasons
                     if variation:
                         return Decision(experiment, variation, enums.DecisionSources.FEATURE_TEST), decide_reasons
 
         # Next check if user is part of a rollout
         if feature.rolloutId:
-            rollout = project_config.get_rollout_from_id(feature.rolloutId)
-            return self.get_variation_for_rollout(project_config, rollout, user_id, attributes)
-        else:
+            return self.get_variation_for_rollout(project_config, feature, user_context, ignore_user_profile)
+
+        # check if not part of experiment
+        if not feature.experimentIds:
+            return Decision(None, None, enums.DecisionSources.FEATURE_TEST), decide_reasons
```

Comment:
Remove this. feature.experimentIds not empty checked in line 523.

Reply:
Done